aten/src/ATen/native/mps/operations/LossOps.mm

```diff
@@ -437,6 +430,20 @@ static void nllnd_loss_forward_impl(Tensor& output,
   if (output.numel() == 0)
     return;
 
+  // https://github.com/pytorch/pytorch/blob/042f2f7746a064f1527d95d1f1d712b4f0b34186/aten/src/ATen/native/cuda/Loss.cu#L335-L346
+  if (target_arg.numel() == 0) {
+    // Here target (and input) have zero elements
+    // Mean reduction on empty tensors produces NaN. See the discussion in
+    // https://github.com/pytorch/pytorch/pull/64572#issuecomment-926504162
+    if (reduction == Reduction::Mean) {
+      output.fill_(std::numeric_limits<double>::quiet_NaN());
```

Comment:
Hmm, wouldn't that need to be a specific to a dtype? Or PyTorch will auto-cast quiet_NaN to float, half and bf16?

Reply:
It's copied from the CUDA code linked on line 433. I'll see if I can add additional tests for this in MPS and amend if necessary.

Reply:
Did a test with

```diff
diff --git a/aten/src/ATen/native/mps/operations/LossOps.mm b/aten/src/ATen/native/mps/operations/LossOps.mm
index c45054f8bc1..31480ffe0a5 100644
--- a/aten/src/ATen/native/mps/operations/LossOps.mm
+++ b/aten/src/ATen/native/mps/operations/LossOps.mm
@@ -443,6 +443,7 @@ static void nllnd_loss_forward_impl(Tensor& output,
     // Mean reduction on empty tensors produces NaN. See the discussion in
     // https://github.com/pytorch/pytorch/pull/64572#issuecomment-926504162
     if (reduction == Reduction::Mean) {
+      NSLog(@"In if (reduction == Reduction::Mean) {");
       output.fill_(std::numeric_limits<double>::quiet_NaN());
     } else {
       output.zero_();
```

```python
import torch
import torch.nn.functional as F

input_size = [2, 3, 0, 7]
reduction = "mean"
dtypes = [torch.float16, torch.float32, torch.bfloat16]

for dtype in dtypes:
    input_cpu = torch.rand(input_size, requires_grad=True, device='cpu', dtype=dtype)
    input_mps = input_cpu.to('mps')

    num_channels = input_size[1]
    target_size = (input_size[0], ) + tuple(input_size[2:])
    target_cpu = torch.randint(num_channels, target_size, device='cpu')
    target_mps = target_cpu.to('mps')

    output_cpu = F.nll_loss(input_cpu, target_cpu, reduction=reduction)
    output_mps = F.nll_loss(input_mps, target_mps, reduction=reduction)
    print(f"{output_mps = }")
    print(f"{torch.isnan(output_cpu) == torch.isnan(output_mps) = }")
```

yields

```
2024-09-10 01:07:25.691 python3.12[94805:555970] In if (reduction == Reduction::Mean) {
output_mps = tensor(nan, device='mps:0', dtype=torch.float16, grad_fn=<NllLoss2DBackward0>)
torch.isnan(output_cpu) == torch.isnan(output_mps) = tensor(True, device='mps:0')
2024-09-10 01:07:25.760 python3.12[94805:555970] In if (reduction == Reduction::Mean) {
output_mps = tensor(nan, device='mps:0', grad_fn=<NllLoss2DBackward0>)
torch.isnan(output_cpu) == torch.isnan(output_mps) = tensor(True, device='mps:0')
2024-09-10 01:07:25.767 python3.12[94805:555970] In if (reduction == Reduction::Mean) {
output_mps = tensor(nan, device='mps:0', dtype=torch.bfloat16, grad_fn=<NllLoss2DBackward0>)
torch.isnan(output_cpu) == torch.isnan(output_mps) = tensor(True, device='mps:0')
```

So looks like autocast does the right thing here. I'll add a test for it in `test_mps.py`. This could probably be extended to `test_nn.py` later.

Reply:
Actually, it looks easy to extend in `test_nn.py` directly, so that seems like a better approach.

Reply:
Done.

Reply:
Assuming we prefer `test_nn.py` when feasible. If the tests should go in `test_mps.py` instead please unresolve.